peps/pep-0734.rst

```diff
@@ -0,0 +1,396 @@
+PEP: 734
+Title: Multiple Interpreters in the Stdlib
+Author: Eric Snow <ericsnowcurrently@gmail.com>
+Status: Draft
+Type: Standards Track
+Created: 06-Nov-2023
+Python-Version: 3.13
+Replaces: 554
+
+
+.. note::
+   This PEP is essentially a continuation of :pep:`554`.  That document
+   had grown a lot of ancillary information across 7 years of discussion.
+   This PEP is a reduction back to the essential information.  Much of
+   that extra information is still valid and useful, just not in the
+   immediate context of the specific proposal here.
+
+Abstract
+========
+
+I propose that we add a new module, ``interpreters``, to the standard
+library, to make the existing multiple-interpreters feature of CPython
+more easily accessible to Python code.  This is particularly relevant
+now that we have a per-interpreter GIL (:pep:`684`) and people are
+more interested in using multiple interpreters.  Without a stdlib
+module, users are limited to the
+:ref:`C API <python:sub-interpreter-support>`, which restricts how much
+they can try out and take advantage of multiple interpreters.
+
+
+Rationale
+=========
+
+The ``interpreters`` module will provide a high-level interface to the
+multiple interpreter functionality.  Since we have no experience with
+how users will make use of multiple interpreters in Python code, we are
+purposefully keeping the initial API as lean and minimal as possible.
+The objective is to provide a well-considered foundation on which we may
+add further (more advanced) functionality later.
+
+That said, the proposed design incorporates lessons learned from
+existing use of subinterpreters by the community, from existing stdlib
+modules, and from other programming languages.  It also factors in
+experience from using subinterpreters in the CPython test suite and
+using them in `concurrency benchmarks`_.
+
+.. _concurrency benchmarks:
+   https://github.com/ericsnowcurrently/concurrency-benchmarks
+
+The module will include a basic mechanism for communicating between
+interpreters.  Without one, multiple interpreters are a much less
+useful feature.
+
+
+Specification
+=============
+
+The module will:
+
+* expose the existing multiple interpreter support
+* introduce a basic mechanism for communicating between interpreters
+
+The module will wrap a new low-level ``_interpreters`` module
+(in the same way as the ``threading`` module).  However, that low-level
+API is not intended for public use and thus not part of this proposal.
+
+We also expect that an ``InterpreterPoolExecutor`` will be added to the
+``concurrent.futures`` module, but that is outside the scope of this PEP.
```

Comment:
I don't mind pulling it in.

Reply:
Great. In the previous PEP version, this didn't follow the Executor model very closely -- instead of a function and optional arguments you were passing a string (like `Interpreter.exec()`) IIUC `ProcessPoolExecutor` has solved this issue, so hopefully `InterpreterPoolExecutor` can solve it similarly. (I'm guessing using `pickle`? Though I've never looked.)

Reply:
Yeah, it would be pickle.